### DISCUSSION TITLE
Graduation / changing protocols

### DISCUSSION THREAD

**Discussion by cxrodgers** [wehr-lab/autopilot#178] [General]
Is there an easy way to change the protocol.json that a Subject is using, without defining graduation criteria? The mouse hasn't actually reached any criterion, I just want to transition the same Subject to a new protocol that I created. In the past I have always just done this by creating a new Subject, but in this case I'd like to keep everything in the same HDF5 file if possible.

Also, on a related note, I see that "reward" is typically a parameter in the protocol.json. Is there a way to modify the reward value (or really any parameter) separately for each mouse? Again, the only way I've done this before is to create a separate protocol for each mouse, but this can lead to bugs because some parameters are supposed to be the same across mice and others are supposed to vary across mice. 

thanks!

**Comment by sneakers-the-rat** [comment]
Yes! 
- The Subject class should automatically detect changes to the .json protocol it is assigned: https://github.com/auto-pi-lot/autopilot/blob/04b5968ba02c8a1413a27eb6a138b6a186b130f1/autopilot/data/subject.py#L538-L548
- The Terminal GUI has an action to batch reassign protocols (Tools > Batch Reassign Protocols) (https://github.com/auto-pi-lot/autopilot/blob/04b5968ba02c8a1413a27eb6a138b6a186b130f1/autopilot/gui/menus/tools.py#L83 and https://github.com/auto-pi-lot/autopilot/blob/04b5968ba02c8a1413a27eb6a138b6a186b130f1/autopilot/agents/terminal.py#L797 )
- Individual Subjects can have their protocols reassigned directly from their `protocol` attribute ( https://docs.auto-pi-lot.com/en/latest/data/subject.html#autopilot.data.subject.Subject.protocol ) using a `Protocol_Status` object ( https://docs.auto-pi-lot.com/en/latest/data/models/subject.html#autopilot.data.models.subject.Protocol_Status ) or the `assign_protocol` method: https://docs.auto-pi-lot.com/en/latest/data/subject.html#autopilot.data.subject.Subject.assign_protocol

Parameters that vary across mice are a great idea, but not currently formally supported. This is something I had intended to write alongside the currently hardcoded pre/post-run dialogues for inputting weight. I think there are at least three types of task parameters that should be supported:

- Task-global parameters that are true for all subjects running a task, to be declared at protocol creation
- Subject-level parameters that vary for each subject running a task, to be declared at protocol assignment
- Session-level parameters that vary for each session, to be declared at start/stop of running a task.

I issued v0.5.0 as an alpha release because there are still a few places that need to be formalized in the new data modeling system, and task parameters are the major one. I'd see this being implemented like this:

- There's a stub for a `Task_Params` class ( https://docs.auto-pi-lot.com/en/latest/data/models/protocol.html#autopilot.data.models.protocol.Task_Params ) that would replace the `PARAMS` dictionary in a Task definition that should either
    - Support Field annotations to declare different parameter types, (which i favor for maintainability and less boilerplate) eg:
    ```python
    class PARAMS(Task_Params):
        global_param: int = Field(description="Some task global param, the default type")
        subject_param: int = Field(
            description="Subject-specific param",
            param_scope="subject")
        session_param:int = Field(
            description="Param given every session",
            param_scope="session")
    ```
    - Require sub-models that carry each of the different scopes, eg:
    ```python
    class Global_Params(Params):
        global_param: int
    
    class Subject_Params(Params):
        subject_param: int
    
    class Session_Params(Params):
        session_param: int
    
    class PARAMS(Task_Params):
        global = Global_Params
        subject = Subject_Params
        session = Session_Params
    ```
- Store these separate classes of parameters in separate attribute dictionaries in the relevant protocol node (see the `.protocol` setter method)
- Be exposed to editing from the GUI (using the model GUI widget: https://docs.auto-pi-lot.com/en/latest/gui/widgets/model.html ) and the subject class ( no extra code, just reassigning the `protocol` property)

Would that cover what you have in mind?

**Reply by sneakers-the-rat** [reply]
There's also a very in-progress effort to make the subject data model general and modifiable from plugins (https://github.com/auto-pi-lot/autopilot/blob/04b5968ba02c8a1413a27eb6a138b6a186b130f1/autopilot/data/models/subject.py#L121-L162), and again an unpopulated stub module for making links between data model fields (`autopilot.data.modeling.links`) so that it would be possible to do this without explicit support, so you would be able to say the `reward` parameter is linked to some subject-specific attribute node in its hdf5 file  like `reward: units.mL = Link('root.task_info.reward')`.

I had also intended to create a `Reward_Manager` class (like the very badly implemented `Stim_Manager`) that could let rewards be managed as the result of a subject's history in the task: eg. to specify that a subject should get a total amount of reward per session, etc. and that would also be much much easier to write now with a formal data modeling system. 

would love to help write/advise/etc. whatever would be the most useful thing to you

**Comment by cxrodgers** [comment]
"Batch Reassign Protocols" is great!! I can't believe I never noticed that before. Everyone here is excited about how this will things easier for us. 

I like the idea of the global/subject/session params. I'll think more about how this would work.
